src/sources/datadog/agent.rs

```diff
@@ -322,10 +487,103 @@ fn decode(header: &Option<String>, mut body: Bytes) -> Result<Bytes, ErrorMessag
             }
         }
     }
-
+    emit!(&DatadogAgentRequestReceived {
+        byte_size: body.len(),
+    });
     Ok(body)
 }
 
+fn into_vector_metric(dd_metric: DatadogSeriesMetric, api_key: Option<Arc<str>>) -> Vec<Event> {
+    let mut tags = BTreeMap::<String, String>::new();
+    for tag in dd_metric.tags.clone().unwrap_or_default() {
+        let kv = tag.split_once(":").unwrap_or((&tag, ""));
+        tags.insert(kv.0.trim().into(), kv.1.trim().into());
+    }
+    dd_metric
+        .host
+        .as_ref()
+        .and_then(|host| tags.insert("host".into(), host.to_owned()));
+    dd_metric
+        .source_type_name
+        .as_ref()
+        .and_then(|source| tags.insert("source_type_name".into(), source.to_owned()));
+    dd_metric
+        .device
+        .as_ref()
+        .and_then(|dev| tags.insert("device".into(), dev.to_owned()));
+
+    match dd_metric.r#type {
+        DatadogMetricType::Count => dd_metric
+            .points
+            .iter()
+            .map(|dd_point| {
+                Metric::new(
+                    dd_metric.metric.clone(),
+                    MetricKind::Incremental,
+                    MetricValue::Counter { value: dd_point.1 },
+                )
+                .with_timestamp(Some(Utc.timestamp(dd_point.0, 0)))
+                .with_tags(Some(tags.clone()))
+            })
+            .map(|mut metric| {
+                if let Some(k) = &api_key {
+                    metric
+                        .metadata_mut()
+                        .set_datadog_api_key(Some(Arc::clone(k)));
+                }
+                metric.into()
+            })
```

Comment:
I think you could move those to a single section after this whole match block, which would let you clean up a bunch of the duplication going on here.